<a href="https://colab.research.google.com/github/pmontman/tmp_choicemodels/blob/main/nb/tutorials/practice_final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Types of questions for the final exam

We will see:
 * Nested logit
 * Ordered logit
 * Panel data and Mixed logit



---
---

# Preparing the environment
*The preparation and dataset loading code is given to the students*

In [1]:
!pip install biogeme

Load the packages, feel free to change the names.

In [2]:
import pandas  as pd
import numpy as np
import matplotlib.pyplot as plt

import biogeme.database as db
import biogeme.biogeme as bio
import biogeme.models as models
import biogeme.expressions as exp
import biogeme.tools as tools
import biogeme.distributions as dist

# Load the dataset

In [3]:
path = 'https://raw.githubusercontent.com/pmontman/pub-choicemodels/main/data/fishing.csv'
data_pd = pd.read_csv(path)

A simple look at the dataset.

In [4]:
data_pd.head(5)

mode  price_beach  price_pier  ...  catch_boat  catch_charter     income
0     4      157.930     157.930  ...      0.2601         0.5391  7083.3317
1     4       15.114      15.114  ...      0.1574         0.4671  1249.9998
2     3      161.874     161.874  ...      0.2413         1.0266  3749.9999
3     2       15.134      15.134  ...      0.1643         0.5391  2083.3332
4     3      106.930     106.930  ...      0.1082         0.3240  4583.3320

[5 rows x 10 columns]

---
---

# Auxiliary functions

The first function takes the dictionary of utilities, a pandas dataframe, and the name of the variable that contains the variable with the results of the choice. It returns the biogeme object with the model and the estimated 'results' object (the one we get the values, likelihoods, etc.)
We have added the dictionary with the utilities to the biogeme object, in case we use it later.

In [5]:
def qbus_estimate_bgm(V, pd_df, tgtvar_name, modelname='bgmdef'):
 av_auto = V.copy()
 for key, value in av_auto.items():
   av_auto[key] = 1
 bgm_db = db.Database(modelname + '_db', pd_df)
 globals().update(bgm_db.variables)
 logprob = models.loglogit (V , av_auto , bgm_db.variables[tgtvar_name] )
 bgm_model = bio.BIOGEME ( bgm_db, logprob )
 bgm_model.utility_dic = V.copy()
 return bgm_model, bgm_model.estimate()

The next function will calculate the predictions for a given biogeme object that was estimated with `qbus_estimate_bgm`. The output is the array with the choice probabilities. From the choice probabilities, this can be used to calculate accuracies, confusion matrices and the output of what-if scenarios.

In [6]:
def qbus_simulate_bgm(qbus_bgm_model, betas, pred_pd_df):
  av_auto = None
  targets = None
  if hasattr(qbus_bgm_model, 'ord_probs'):
    av_auto = qbus_bgm_model.ord_probs.copy()
    targets = qbus_bgm_model.ord_probs.copy()
  else:
    av_auto = qbus_bgm_model.utility_dic.copy()
    targets = qbus_bgm_model.utility_dic.copy()

  for key, value in av_auto.items():
    av_auto[key] = 1



  for key, value in targets.items():
    if hasattr(qbus_bgm_model, 'nest_tuple'):
      targets[key] = models.nested(qbus_bgm_model.utility_dic, av_auto, qbus_bgm_model.nest_tuple, key)
    else:
      if hasattr(qbus_bgm_model, 'ord_probs'):
       0
       #targets[key] = qbus_bgm_model.ord_probs[key]
      else:
       targets[key] = models.logit(qbus_bgm_model.utility_dic, av_auto, key)

  bgm_db = db.Database('simul', pred_pd_df)
  globals().update(bgm_db.variables)
  bgm_pred_model = bio.BIOGEME(bgm_db, targets)
  simulatedValues = bgm_pred_model.simulate(betas)
  return simulatedValues

The function `qbus_calc_accu_confusion` calculates the accuracies given the choice probability predictions a pandas dataset and the specification of the name that contains the actual choices in the input dataset.

In [7]:
def qbus_calc_accu_confusion(sim_probs, pd_df, choice_var):
  which_max = sim_probs.idxmax(axis=1)
  data = {'y_Actual':   pd_df[choice_var],
          'y_Predicted': which_max
        }

  df = pd.DataFrame(data, columns=['y_Actual','y_Predicted'])
  confusion_matrix = pd.crosstab(df['y_Actual'], df['y_Predicted'], rownames=['Actual'], colnames=['Predicted'])
  accu = np.mean(which_max == pd_df[choice_var])
  return accu, confusion_matrix 

The next function calculates the likelihood ratio test having to write a bit less code that the default biogeme function. The arguments are the results objects of the two models to be compared. The first is the more complex and the second is the reference model (**the order is important!**). The third argument is the significance level for the test.

In [8]:
def qbus_likeli_ratio_test_bgm(results_complex, results_reference, signif_level):
  return tools.likelihood_ratio_test( (results_complex.data.logLike, results_complex.data.nparam),
                                     (results_reference.data.logLike, results_reference.data.nparam), signif_level)

The next function just updates the globals so we can use it 

In [9]:
def qbus_update_globals_bgm(pd_df):
   globals().update(db.Database('tmp_bg_bgm_for_glob', pd_df).variables)

The next function calculates the nested logit version. Similar to the multinomial logit

In [10]:
def qbus_estimate_nested_bgm(V, pd_df, nests,  tgtvar_name, modelname='bgmdef'):
 av_auto = V.copy()
 for key, value in av_auto.items():
   av_auto[key] = 1
 bgm_db = db.Database(modelname + '_db', pd_df)
 globals().update(bgm_db.variables)
 logprobnest = models.lognested (V, av_auto , nests , bgm_db.variables[tgtvar_name] )
 #logprob = models.loglogit (V , av_auto , bgm_db.variables[tgtvar_name] )
 bgm_model = bio.BIOGEME ( bgm_db, logprobnest )
 bgm_model.utility_dic = V.copy()
 bgm_model.nest_tuple = nests
 return bgm_model, bgm_model.estimate()

The auxiliary function for the ordered logit. The use is slightly different from the basic multinomial logit!
* The `V` argument is just the expression of a utility function, not the dictionary mapping alternative ids to the utility functions.
* The argument `ord_alt_ids` is a list with the ids of the alternatives **in the order that we want to impose**.The parameter to know about.

Then the rest of the arguments are used as usual `pd_df` the pandas dataframe, `tgt_varname` the name of the variable with the choices, and an optional `modelname`.

In [11]:
def qbus_estimate_ordered_bgm(V, ord_alt_ids, pd_df, tgtvar_name, modelname='ord_bgm'):
 bgm_db = db.Database(modelname + '_db', pd_df)
 globals().update(bgm_db.variables)

 taus_map = {ord_alt_ids[0]: exp.Beta('tau1', -1, None, None, 0) }
 i = 1
 for id in ord_alt_ids[1:-1]:
  taus_map[id] = taus_map[ ord_alt_ids[i-1] ] + exp.Beta('delta_'+ str(i + 1), i, 0, None, 0)
  i = i + 1

 alt_probs_map = {ord_alt_ids[0]: dist.logisticcdf( taus_map[ord_alt_ids[0] ] - V_ord) }

 i = 1
 for id in ord_alt_ids[1:-1]:
  alt_probs_map[id] = dist.logisticcdf( taus_map[id] - V_ord) - dist.logisticcdf( taus_map[ ord_alt_ids[i-1] ] - V_ord)
  i = i + 1

 alt_probs_map[ord_alt_ids[i] ] = 1 - dist.logisticcdf( taus_map[ord_alt_ids[i-1]] - V_ord)

 logprob = exp.log(exp.Elem(alt_probs_map, bgm_db.variables[tgtvar_name]))

 #logprob = models.loglogit (V , av_auto , bgm_db.variables[tgtvar_name] )
 bgm_model = bio.BIOGEME ( bgm_db, logprob )
 bgm_model.utility_dic = V
 bgm_model.ord_probs = alt_probs_map.copy()
 return bgm_model, bgm_model.estimate()

The mixed logit with panel data



In [12]:
def qbus_estimate_mixed_bgm(V, pd_df, tgtvar_name, panelvar_name=None, n_draws=50, seed=1, modelname='bgmdef'):
 do_panel = not (panelvar_name==None)

 av_auto = V.copy()
 for key, value in av_auto.items():
   av_auto[key] = 1
 bgm_db = db.Database(modelname + '_db', pd_df)
 if (do_panel):
   bgm_db.panel(panelvar_name)

 globals().update(bgm_db.variables)
 #logprob = models.loglogit (V , av_auto , bgm_db.variables[tgtvar_name] )
 obsprob = models.logit(V, av_auto, CHOICE)
 if (do_panel):
  condprobIndiv = exp.PanelLikelihoodTrajectory(obsprob)
 else:
  condprobIndiv = obsprob
 logprob = exp.log(exp.MonteCarlo(condprobIndiv))
 bgm_model  = bio.BIOGEME(bgm_db,logprob,numberOfDraws=n_draws, seed=seed)
 bgm_model.utility_dic = V.copy()
 return bgm_model, bgm_model.estimate()

---
---

# Estimating a baseline model

In [13]:
ASC_beach = exp.Beta ( 'ASC_beach' ,0, None , None ,0)
ASC_pier = exp.Beta ( 'ASC_pier' ,0, None , None ,0)
ASC_boat = exp.Beta ( 'ASC_boat' ,0, None , None ,0)
ASC_charter = exp.Beta ( 'ASC_charter' ,0, None , None ,1)
B_price = exp.Beta ( 'B_price' ,0, None , None ,0)
B_catch = exp.Beta ( 'B_catch' ,0, None , None ,0)

In [14]:
qbus_update_globals_bgm(data_pd)

In [15]:
V_beach = ASC_beach + B_price*price_beach + B_catch*catch_beach
V_pier = ASC_pier + B_price*price_pier + B_catch*catch_pier
V_boat = ASC_boat + B_price*price_boat + B_catch*catch_boat
V_charter = ASC_charter + B_price*price_charter + B_catch*catch_charter

V_base = {1: V_beach,
     2: V_pier,
     3: V_boat,
     4: V_charter}

In [16]:
model_base, results_base = qbus_estimate_bgm(V_base, data_pd, 'mode', 'fish')

In [17]:
results_base.getEstimatedParameters()

Value   Std err  ...  Rob. t-test  Rob. p-value
ASC_beach -1.498836  0.132932  ...   -11.555999  0.000000e+00
ASC_boat  -0.627457  0.100813  ...    -5.370843  7.836923e-08
ASC_pier  -1.191788  0.128483  ...    -9.287130  0.000000e+00
B_catch    0.377234  0.109971  ...     3.163442  1.559155e-03
B_price   -0.024789  0.001704  ...   -10.645256  0.000000e+00

[5 rows x 7 columns]

---
---

# Nested logit Question: Estimate the model 
Nested logit, we declare the nest parameter, the parameter that will represent the 'strength' of the nesting.

We use the biogeme expression (same as for the Beta coefficients). In biogeme
the larger the parameter, the stronger the nesting. Its minimum value should be 1, the maximum can be a large enough number such as 100. This comes from the $\lambda$ in the theory part that is between 0 and 1, biogeme use a $\mu = 1/ \lambda$ so it should be a number larger than 1 if $\lambda$ is between 0 and 1.

In [18]:
MU_nest_A = exp.Beta('MU_nest_A', 1, 1, 100, 0)

And the nest structure. In biogeme, this is done via a python tuple of two elements:
1. The first element is the the parameter
2. The second parameter is a python list of ids of the alternatives that are affected by the grouping. **Remember to map the ids to the alternatives correctly!**

In the following cell, we create a grouping for beach (alt. 1), pier (alt. 2) and private boat (alt. 3). This is the grouping that will be affected by the `MU_NONCHART`. Other group, in this case it is only formed by the charter boat.
Since there is only one altenative in the group, we can set its grouping parameter to 1.
Finaly we put all tuples together to create the full grouping specification that is passed to biogeme.

In [19]:
nest_A = MU_nest_A, [1, 2, 3]
nest_B = 1.0, [4]
nests_struct = nest_A, nest_B

Then we will estimate and check the model parameters.

In [20]:
model_nest, results_nest = qbus_estimate_nested_bgm(V_base, data_pd, nests_struct, 'mode', 'data_nest' )

The estimation of the model gives similar coefficients to the paper.

In [21]:
results_nest.getEstimatedParameters()

Value   Std err  ...  Rob. t-test  Rob. p-value
ASC_beach -0.391672  0.183913  ...    -1.865064  6.217248e-02
ASC_boat  -0.063007  0.130176  ...    -0.435367  6.632961e-01
ASC_pier  -0.241545  0.166548  ...    -1.278783  2.009735e-01
B_catch    0.414475  0.101409  ...     4.219833  2.444838e-05
B_price   -0.014047  0.002188  ...    -5.909232  3.437056e-09
MU_nest_A  2.280453  0.395305  ...     4.613708  3.955487e-06

[6 rows x 7 columns]

To check is the nested model is a statistically significant improvement over the multinomial logit version, we can do a likelihood ratio test, as usual, when the complex model is nested. If the only thing that changes is the inclusion of the 'nesting parameter' we test if nesting is useful. We can check several modifications at once, but then we would not be able to sepearate the nesting part form the rest. 

In [22]:
qbus_likeli_ratio_test_bgm(results_nest, results_base, 0.95)

('H0 can be rejected at level 0.95', 27.929484121169025, 3.841458820694124)

---
---

# Nested logit question: Find/Discuss the nest structure in the data

Apart from knowing how to estimate the nested logit and interpret the results, we need to be able to compare different nesting structures that might be applicable to the data, to pick the one that is better. In practice, sometimes there are several 'reasonable' nesting structures .

In [23]:
MU_nest_B = exp.Beta('MU_nest_B', 1, 1, 100, 0)

In [24]:
alt_nest_A = MU_nest_A, [1, 2, 4]
alt_nest_B = MU_nest_B, [3]
alt_nests_struct = alt_nest_A, alt_nest_B

In [25]:
alt_model_nest, alt_results_nest = qbus_estimate_nested_bgm(V_base, data_pd, alt_nests_struct, 'mode', 'data_nest' )

In [26]:
alt_results_nest.getEstimatedParameters()

Value  Active bound  ...   Rob. t-test  Rob. p-value
ASC_beach -1.234321           0.0  ... -3.997927e+00  6.389967e-05
ASC_boat  -0.602359           0.0  ... -5.435830e+00  5.454193e-08
ASC_pier  -0.976813           0.0  ... -3.760600e+00  1.695062e-04
B_catch    0.375114           0.0  ...  3.378004e+00  7.301393e-04
B_price   -0.021707           0.0  ... -5.061436e+00  4.161101e-07
MU_nest_A  1.209235           0.0  ...  4.331897e+00  1.478299e-05
MU_nest_B  1.000000           1.0  ...  3.050300e+18  0.000000e+00

[7 rows x 8 columns]

What happens when comparing the nested logit is that we cannot always use the likelihood ratio test, because the models to compare are not always nested (one is not always is included in the other). When models are very similar, just differing on the nested structure and the same number of parameters, we can compare the likelihoods directly (the larger the better). This will not give us a nice p-value but it can help decide among nest structures.

**However,** it more general if we just use the  Akaike information criteria directly (AIC). Remember that lower means a better model, and in practice is very conservative (adding parameter has to improve the model by 'a lot')..

In [27]:
#Likelihood ratio test is rarely applicable to compare between nests
#qbus_likeli_ratio_test_bgm(alt_results_nest, results_nest, 0.95)

In [28]:
#alt_results_nest.data.logLike

In [29]:
#results_nest.data.logLike

We should go directly for the Akaike information criteria (the lower the better)..

In [30]:
alt_results_nest.data.akaike

2474.0549750198943

In [31]:
results_nest.data.akaike

2445.6381771493216

The results from loglik and Akaike are consistent in this case

In [32]:
results_base.data.akaike

2471.5676612704906

---
---

# Ordered Logit question: Estimate the model and interpretation

The ordered logit considers only one utility function, not one per alternative.
The main 'challenge' in the ordered logit is realizing that the problem we are dealing with can be modeled with the ordered logit, that the alternatives have some sort of 'natural' order.

In [33]:
ASC_ord = exp.Beta('ASC_ord', 0, None, None, 0)

In [34]:
V_ord = ASC_ord + B_price*price_beach + B_catch*catch_beach + B_price*price_pier + B_catch*catch_pier

We use the auxiliary function `qbus_estimate_ordered_bgm`, notice the argument `[1,3,2,4]` specifying the desired order of the alternatives. In this case, it means we think that the order is:

 beach < private boat < pier < charterboat.
 
 **Please pay attention to the order that we specify! Otherwise we are incorrectly specify the model.**



In [35]:
qord_model, qord_results = qbus_estimate_ordered_bgm(V_ord, [1,3,2,4], data_pd, 'mode', )

In [36]:
qord_results.getEstimatedParameters()

Value   Std err     t-test  ...  Rob. Std err  Rob. t-test  Rob. p-value
ASC_ord  0.455434  0.058527   7.781605  ...      0.060283     7.554879  4.196643e-14
B_catch -0.046664  0.167358  -0.278824  ...      0.174750    -0.267031  7.894457e-01
B_price  0.000794  0.000274   2.893848  ...      0.000289     2.744581  6.058818e-03
delta_2  1.921617  0.087904  21.860342  ...      0.087542    21.950819  0.000000e+00
delta_3  0.609226  0.042812  14.230393  ...      0.042970    14.177963  0.000000e+00
tau1    -1.455434  0.058527 -24.867737  ...      0.060283   -24.143185  0.000000e+00

[6 rows x 7 columns]

The interpretation of the coefficients is slightly different because there is only one utility, therefoe, positive numbers that increase utility make the alternatives later in our specific order more likely. The interpretation of the cutoff points tau1 would be the fist cutoff, then tau2 would be tau1+delta_2, tau3 = tau2 + delta_3 and so on. Given the expression of the logit, it is difficult to get some clear intuition on the differences here, it is better to look at the predicions/simuations directly.

In [37]:
qbus_simulate_bgm(qord_model, qord_results.getBetaValues(), data_pd)

1         3         2         4
0     0.103755  0.337871  0.150959  0.407416
1     0.126986  0.371450  0.147900  0.353664
2     0.106979  0.343098  0.150739  0.399183
3     0.126966  0.371424  0.147904  0.353707
4     0.111532  0.350151  0.150300  0.388016
...        ...       ...       ...       ...
1177  0.108758  0.345899  0.150585  0.394758
1178  0.096320  0.325042  0.151134  0.427505
1179  0.122585  0.365778  0.148707  0.362930
1180  0.127520  0.372117  0.147796  0.352567
1181  0.080775  0.294361  0.149589  0.475274

[1182 rows x 4 columns]

---
---

# Ordered Logit question: Finding the order

The situation when several possible orders seem 'reasonable' is a bi far-fetched, but it is also simple to test. We can use Akaike.

In [38]:
qord_model_na, qord_results_na = qbus_estimate_ordered_bgm(V_ord, [2,4,3,1], data_pd, 'mode')

In [39]:
qord_results_na.getGeneralStatistics()

{'Akaike Information Criterion': (2985.909279383948, '.7g'),
 'Bayesian Information Criterion': (3016.3590585717443, '.7g'),
 'Excluded observations': (0, ''),
 'Final gradient norm': (0.000955210400516675, '.4E'),
 'Final log likelihood': (-1486.954639691974, '.7g'),
 'Init log likelihood': (-1778.778200033079, '.7g'),
 'Likelihood ratio test for the init. model': (583.6471206822098, '.7g'),
 'Nbr of threads': (2, ''),
 'Number of estimated parameters': (6, ''),
 'Rho-square for the init. model': (0.1640584308575842, '.3g'),
 'Rho-square-bar for the init. model': (0.1606853290285375, '.3g'),
 'Sample size': (1182, '')}

We see the new order is better than the original, but both orders are really bad ideas...

In [40]:
qord_results_na.data.akaike

2985.909279383948

In [41]:
qord_results.data.akaike

2998.539958503067

In [42]:
results_base.data.akaike

2471.5676612704906

---
---


# Panel and mixed logit questions

We do not have auxiliary functions for the mixed model (yet) so we have to estimate it manually. We will go back to the Swissmetro dataset because it shows a nice result.

# Panel and mixed logit questions: Panel vs Nonpanel

What happens when we include the panel information?

Toggle `do_PANEL` and the Runtime -> Run After from here to check the differences.

In [43]:
do_PANEL = True

In [44]:
#path = 'https://raw.githubusercontent.com/pmontman/pub-choicemodels/main/data/catsup.csv'
#data_pd = pd.read_csv(path)
data_pd = pd.read_csv('http://transp-or.epfl.ch/data/swissmetro.dat', sep='\t')

#data_pd

The basic cleanup of the Swissmetro dataset, remove invalid choices (0s in the CHOICE variable)

In [45]:
data_pd = data_pd[ data_pd['CHOICE'] > 0]
data_bgm = db.Database("swiss", data_pd)

In [46]:
globals().update(data_bgm.variables)

In [47]:
if (do_PANEL):
 data_bgm.panel("ID")

In [48]:
ASC_CAR = exp.Beta ( 'ASC_CAR' ,0, None , None ,0)
ASC_TRAIN = exp.Beta ( 'ASC_TRAIN' ,0, None , None ,0)
ASC_SM = exp.Beta ( 'ASC_SM' ,0, None , None ,1)
B_TIME = exp.Beta ( 'B_TIME' ,0, None , None ,0)
B_COST = exp.Beta ( 'B_COST' ,0, None , None ,0)

In [49]:
V1 = ASC_TRAIN + B_TIME * TRAIN_TT + B_COST * TRAIN_CO 
V2 = ASC_SM + B_TIME * SM_TT + B_COST * SM_CO 
V3 = ASC_CAR + B_TIME * CAR_TT + B_COST * CAR_CO

In [50]:
V = {1: V1 ,
2: V2 ,
3: V3 }

In [51]:
av = {1: TRAIN_AV,
2: SM_AV,
3: CAR_AV }

In [52]:
logprob = models.loglogit (V , av , CHOICE )
bgm_model = bio.BIOGEME ( data_bgm, logprob )
results = bgm_model.estimate()

In [53]:
results.getEstimatedParameters()

Value   Std err  ...  Rob. t-test  Rob. p-value
ASC_CAR    0.111463  0.097955  ...     0.734403  4.627033e-01
ASC_TRAIN -1.171662  0.140207  ...    -5.292731  1.205028e-07
B_COST     0.000236  0.000070  ...     3.903901  9.465462e-05
B_TIME    -0.008579  0.001166  ...    -3.579591  3.441329e-04

[4 rows x 7 columns]

---
---

# Mixed logit question: Estimation

In [54]:
SIGMA_B_COST = exp.Beta('SIGMA_B_COST',0,0,None,0)

In [55]:
EC_B_COST = SIGMA_B_COST * exp.bioDraws('EC_B_COST','NORMAL')

In [56]:
V1 = ASC_TRAIN + B_TIME * TRAIN_TT + (B_COST + EC_B_COST)  * TRAIN_CO 
V2 = ASC_SM + B_TIME * SM_TT + (B_COST + EC_B_COST) * SM_CO 
V3 = ASC_CAR + B_TIME * CAR_TT + (B_COST + EC_B_COST) * CAR_CO

In [57]:
V = {1: V1, 2:V2, 3:V3}

In [58]:
obsprob = models.logit(V, av, CHOICE)


In [59]:
if (do_PANEL):
 condprobIndiv = exp.PanelLikelihoodTrajectory(obsprob)
else:
 condprobIndiv = obsprob

In [60]:
#condprobIndiv = exp.PanelLikelihoodTrajectory(obsprob)

And Step 2 we take the model and the modifying it by the expresion `exp.MonteCarlo`. The final log it to take the loglikelihood.

In [61]:
logprob = exp.log(exp.MonteCarlo(condprobIndiv))

We we are using simulation, we have to tell biogeme how many draws from the distribution are we going to generate. The more draws, the more accurate estimation, but it is compuationally costly.

We also set up a seed, so we can get the same results if the run the notebook again (setting up a seed is a good habit in general)

In [62]:

# Create the Biogeme object
biogeme  = bio.BIOGEME(data_bgm,logprob,numberOfDraws=50, seed=1)


In [63]:
results_mixed = biogeme.estimate()

In [64]:
results_mixed.getEstimatedParameters()

Value        Std err  ...    Rob. t-test  Rob. p-value
ASC_CAR       0.027109  1.491668e-155  ...  1.797693e+308           0.0
ASC_TRAIN    -0.721633  1.491668e-155  ...  1.797693e+308           0.0
B_COST       -0.010624  1.491668e-155  ...  1.797693e+308           0.0
B_TIME       -0.016584  1.491668e-155  ...  1.797693e+308           0.0
SIGMA_B_COST  0.019808  1.491668e-155  ...  1.797693e+308           0.0

[5 rows x 7 columns]

---
---

# Mixed Logit question: Interpretation of the distribution of a parameter

* The main interpretation is the understanding that parameters are 'random' so the output is an estimation of the probability distribution of the parameters, in this case COST follows a normal $N(-0.0106, 0.020^2)$.

* Secondary interpretations are doing 'something' with the distribution. 
 For example: **What percentage of people in the population have a cost parameter that is positive?**

In [65]:
from scipy.stats import norm

1 - norm.cdf(0, results_mixed.getBetaValues()['B_COST'], results_mixed.getBetaValues()['SIGMA_B_COST'])

0.295855117919254